irc_client/backend.py

```diff
@@ -189,124 +266,39 @@ def _split_line(line: str) -> _Message:
-                        nicks = self._names_replys.pop(channel)
-                        self.event_queue.put((IrcEvent.self_joined, channel, nicks))
-
-                    else:
-                        # TODO: there must be a better way than relying on MOTD
-                        if msg.command == _RPL_ENDOFMOTD:
-                            for channel in self._autojoin:
-                                self.join_channel(channel)
-
-                        self.event_queue.put(
-                            (IrcEvent.server_message, msg.sender, msg.command, msg.args)
-                        )
-
-                else:
-                    self.event_queue.put(
-                        (IrcEvent.unknown_message, msg.sender, msg.command, msg.args)
-                    )
-
-            elif event == _IrcInternalEvent.should_join:
-                [channel] = args
-                self._send("JOIN", channel)
-
-            elif event == _IrcInternalEvent.should_part:
-                channel, reason = args
-                if reason is None:
-                    self._send("PART", channel)
-                else:
-                    # FIXME: the reason thing doesn't seem to work
-                    self._send("PART", channel, ":" + reason)
-
-            elif event == _IrcInternalEvent.should_quit:
-                assert not args
-                self._send("QUIT")
-                self._running = False
-
-            elif event == _IrcInternalEvent.should_send_privmsg:
-                recipient, text = args
-                self._send("PRIVMSG", recipient, ":" + text)
-                self.event_queue.put((IrcEvent.sent_privmsg, recipient, text))
-
-            elif event == _IrcInternalEvent.should_change_nick:
-                [new_nick] = args
-                self._send("NICK", new_nick)
+        return _ReceivedAndParsedMessage(sender, sender_is_server, command, args)
 
-            else:
-                raise RuntimeError("Unrecognized internal event!")
+    def _recv_loop(self) -> None:
+        try:
+            # TODO: does quitting work in all cases, even during connecting?
+            while self._running:
+                line = self._recv_line()
+                if not line:
+                    # "Empty messages are silently ignored"
+                    # https://tools.ietf.org/html/rfc2812#section-2.3.1
+                    continue
+                if line.startswith("PING"):
+                    self._send_soon(line.replace("PING", "PONG", 1))
+                    continue
 
-            self._internal_queue.task_done()
+                message = self._split_line(line)
+                try:
+                    self._handle_received_message(message)
+                except Exception:
+                    traceback.print_exc()
+        finally:
+            assert self._sock is not None
+            self._sock.close()
+            self._sock = None
+            self.event_queue.put(SelfQuit())
```

Comment:
It now goes to event queue in two places. Apparently server is supposed to respond "ok you can quit now"?
```suggestion
```